pkg/openstack/dataplane.go

```diff
@@ -0,0 +1,75 @@
+package openstack
+
+import (
+	"context"
+
+	"github.com/openstack-k8s-operators/lib-common/modules/common/helper"
+	corev1beta1 "github.com/openstack-k8s-operators/openstack-operator/apis/core/v1beta1"
+
+	//metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	//"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	//"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+	dataplanev1 "github.com/openstack-k8s-operators/dataplane-operator/api/v1beta1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+// GetDataplaneNodesets - returns the dataplanenodesets in the namespace of the controlplane
+func GetDataplaneNodesets(ctx context.Context, instance *corev1beta1.OpenStackControlPlane, helper *helper.Helper) (ctrl.Result, *dataplanev1.OpenStackDataPlaneNodeSetList, error) {
+	// Get the dataplane nodesets
+	dataplaneNodesets := &dataplanev1.OpenStackDataPlaneNodeSetList{}
+	opts := []client.ListOption{
+		client.InNamespace(instance.Namespace),
+	}
+	err := helper.GetClient().List(ctx, dataplaneNodesets, opts...)
+	if err != nil {
+		return ctrl.Result{}, nil, err
+	}
+	return ctrl.Result{}, dataplaneNodesets, nil
+}
+
+func DataplaneNodesetsDeployedVersionIsSet(dataplaneNodesets *dataplanev1.OpenStackDataPlaneNodeSetList) bool {
+	for _, nodeset := range dataplaneNodesets.Items {
+		// FIXME: DeployedVersion on the DataplaneNodeset should be a string pointer to match how Controlplane implements this
+		if nodeset.Status.DeployedVersion == "" {
+			return false
+		}
+	}
+	return true
+}
+
+//if !openstack.OVNControllerImageCheck(controlPlane, instance) ||
+//!controlPlane.Status.Conditions.IsTrue(corev1beta1.OpenStackControlPlaneOVNReadyCondition)
+
+// returns true if OVNControllers are deployed on all nodesets
+func DataplaneNodesetsOVNControllerDeployedCheck(version *corev1beta1.OpenStackVersion, dataplaneNodesets *dataplanev1.OpenStackDataPlaneNodeSetList) bool {
+	for _, nodeset := range dataplaneNodesets.Items {
+		if nodeset.Generation != nodeset.Status.ObservedGeneration {
+			return false
+		}
+		if !nodeset.IsReady() {
+			return false
+		}
+		if nodeset.Status.ContainerImages["OvnControllerImage"] != "" && nodeset.Status.ContainerImages["OvnControllerImage"] != *version.Status.ContainerImages.OvnControllerImage {
```

Comment:
If `nodeset.Status.ContainerImages["OvnControllerImage"]` is current `""`, wouldn't that mean we should `return false` for `"returns true if OVNControllers are deployed on all nodesets"`?  Because if there's no image, then by inference OVNController is not deployed?

Reply:
so the logic here is different than everywhere else in openstack-operator because dataplane-operator is using strings instead of *string for its containerImage parameters. But my intent here was to ensure it was set to something before testing it. Also, as far as I know it would be possible to have a DataplaneNodeSet that doesn't deploy a specific service and thus we may need a fall through for that case so the minor update could proceed.

Reply:
> Also, as far as I know it would be possible to have a DataplaneNodeSet that doesn't deploy a specific service and thus we may need a fall through for that case so the minor update could proceed.

If there are NodeSets without OVN controller, then I think we don't need to wait for the readyness of those NodeSets either.

Reply:
should we label NodeSets with OVN then so we can more easily code this. I guess I can sort through them all here just as well but a label might be slightly cleaner.

Reply:
good catches here. I have adjusted the logic. See what you think now

Reply:
looks good to me